lightning/src/ln/channelmanager.rs

```diff
@@ -3321,6 +3377,100 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			self.finish_force_close_channel(failure);
 		}
 	}
+
+	fn set_payment_hash_secret_map(&self, payment_hash: PaymentHash, payment_preimage: Option<PaymentPreimage>, min_value_msat: Option<u64>, invoice_expiry_delta_secs: u32, user_payment_id: u64) -> Result<PaymentSecret, APIError> {
+		assert!(invoice_expiry_delta_secs <= 60*60*24*365); // Sadly bitcoin timestamps are u32s, so panic before 2106
+
+		let payment_secret = PaymentSecret(self.keys_manager.get_secure_random_bytes());
+
+		let _persistence_guard = PersistenceNotifierGuard::new(&self.total_consistency_lock, &self.persistence_notifier);
+		let mut payment_secrets = self.pending_inbound_payments.lock().unwrap();
+		match payment_secrets.entry(payment_hash) {
+			hash_map::Entry::Vacant(e) => {
+				e.insert(PendingInboundPayment {
+					payment_secret, min_value_msat, user_payment_id, payment_preimage,
+					// We assume that last_node_announcement_serial is pretty close to the current
+					// time - its updated when we receive a new block with the maximum time we've
+					// seen in a header, and only incremented by one when we create a node
+					// announcement. As long as we aren't creating a new node announcement more
+					// often than once per second, it should never be more than two hours in the
+					// future. Thus, we add two hours here as a buffer to ensure we absolutely
+					// never fail a payment too early.
+					// Note that we assume that received blocks have reasonably up-to-date
+					// timestamps.
+					expiry_time: self.last_node_announcement_serial.load(Ordering::Acquire) as u64 + invoice_expiry_delta_secs as u64 + 7200,
```

Comment:
Can a counterparty replay announcement signatures to prevent expiration of our pending payments ?

In `internal_announcement_signatures`, we generate a `BroadcastChannelAnnouncement` if signatures validation is successful. But we don't seem to prevent replay.

Following `BroadcastChannelAnnouncement` documentation it says :

> Note that after doing so, you very likely (unless you did so very recently) want to call ChannelManager::broadcast_node_announcement...

If implemented automatically, this call to `broadcast_node_announcement` will increase our `self.last_node_announcement_serial` by one.

In `best_block_updated`, we'll keep the max between current `last_node_announcement_serial` and header time. If old serial is superior to header time we'll never expire payments ?

I think potential manipulation of `last_node_announcement_serial` is already there before this PR ? Maybe we should upper bound on some assumptions on block variance. Like no more than 300 updates per-block interval.

Reply:
Hmm, good point, I hadn't caught the docs suggesting that (I'd only looked at the explicit uses of `last_node_announcement_serial`. If we feel strongly, we should probably just add a second variable which only tracks header times rather than bothering mixing two different uses in the same variable. As for the general case around `last_node_announcement_serial`, do we actually care if we end up with some super high serial number for our node announcements? Last I checked the spec indicated that's allowed but I may be wrong now.

Reply:
> As for the general case around last_node_announcement_serial, do we actually care if we end up with some super high serial number for our node announcements? Last I checked the spec indicated that's allowed but I may be wrong now.

Yes i think we don't care about a super high serial number for our node announcements, unless other implementations have some custom upper bounds or discard any further announcements about some rate. Though to be honest, accidental or malicious partitions from the rest of the gossip layer are still ranking as a low-priority for me. I was more concerned with the payment reception interference here.

Thanks for adding `highest_seen_timestamp`.